### PR TITLE
Remove dependency from DB::Context in remote/cache readers

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 7478d0b5483e6db4fcb9c2e3e094512698a0e052 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49426/7478d0b5483e6db4fcb9c2e3e094512698a0e052/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by kssenii** [comment]
> Stress test (ubsan) — Killed by signal 

Not related to changes, https://github.com/ClickHouse/ClickHouse/issues/49599

**Comment by kssenii** [comment]
> Integration tests (release) [2/4] — fail: 2

`test_merge_tree_s3/test.py::test_store_cleanup_disk_s3` - flaky in master

> Stateless tests (debug, s3 storage) [6/6] — Tests are not finished

https://github.com/ClickHouse/ClickHouse/issues/49185

> Stateless tests (release, s3 storage) [2/2] — fail: 1, Stateless tests (release, wide parts enabled) — fail: 1

00995_exception_while_insert
01079_parallel_alter_add_drop_column_zookeeper

does not look related

File: src/Disks/IO/getThreadPoolReader.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+IAsynchronousReader & getThreadPoolReader(FilesystemReaderType type)
+{
+#ifndef CLICKHOUSE_PROGRAM_STANDALONE_BUILD
```
Comment (alesapin) [inline src/Disks/IO/getThreadPoolReader.cpp:24]: #ifdef

**Review by alesapin — APPROVED** [review]

**Comment by kssenii** [comment]
> Integration tests (asan) [4/6] — fail: 1 

`test_storage_postgresql/test.py::test_concurrent_queries` - not related to changes

> Stateless tests (ubsan) [2/2] — fail: 1

`01710_projection_aggregation_in_order`

```
DB::Exception: No projection is used when allow_experimental_projection_optimization = 1 and force_optimize_projection = 1.
```

not related to changes

> Stress test (ubsan) — Killed by signal

not related to changes

**Pull request merged**